utils/threadpool/fair_threadpool.cpp

```diff
@@ -0,0 +1,295 @@
+    jobsList->push_back(job);
+    txn2JobsListMap_[job.txnIdx_] = jobsList;
+    WeightT currentTopWeight = weightedTxnsQueue_.empty() ? 0 : weightedTxnsQueue_.top().first;
+    weightedTxnsQueue_.push({currentTopWeight, job.txnIdx_});
+  }
+  else
+  {
+    jobsListMapIter->second->push_back(job);
+  }
+
+  if (useLock)
+    newJob.notify_one();
+}
+
+void FairThreadPool::removeJobs(uint32_t id)
+{
+  boost::mutex::scoped_lock lk(mutex);
+
+  for (auto& txnJobsMapPair : txn2JobsListMap_)
+  {
+    ThreadPoolJobsList* txnJobsList = txnJobsMapPair.second;
+    auto job = txnJobsList->begin();
+    while (job != txnJobsList->end())
+    {
+      if (job->id_ == id)
+      {
+        job = txnJobsList->erase(job); // update the job iter
+        if (txnJobsList->empty())
+        {
+          txn2JobsListMap_.erase(txnJobsMapPair.first);
+          delete txnJobsList;
+          break;
+          // There is no clean-up for PQ. It will happen later in threadFcn
+        }
+        continue; // go-on skiping job iter increment
+      }
+      ++job;
+    }
+  }
+}
+
+void FairThreadPool::threadFcn(const Priority preferredQueue)
+{
+  if (preferredQueue == EXTRA)
+    utils::setThreadName("Extra");
+  else
+    utils::setThreadName("Idle");
+  RunListT runList; // This is a vector to allow to grab multiple jobs
+  RescheduleVecType reschedule;
+  bool running = false;
+  bool rescheduleJob = false;
+
+  try
+  {
+    while (!_stop)
+    {
+      runList.clear(); // remove the job
+      boost::mutex::scoped_lock lk(mutex);
+
+      if (preferredQueue == EXTRA && stopExtra)
+      {
+        --extraThreads;
+        return;
+      }
+
+      if (weightedTxnsQueue_.empty())
+      {
+        newJob.wait(lk);
+        continue; // just go on w/o re-taking the lock
+      }
+
+      WeightedTxnT weightedTxn = weightedTxnsQueue_.top();
+      auto txnAndJobListPair = txn2JobsListMap_.find(weightedTxn.second);
+      // Looking for non-empty jobsList in a loop
+      // Waiting on cond_var if PQ is empty(no jobs in this thread pool)
+      while (txnAndJobListPair == txn2JobsListMap_.end() || txnAndJobListPair->second->empty())
+      {
+        if (txnAndJobListPair != txn2JobsListMap_.end()) // JobList is empty
+        {
+          ThreadPoolJobsList* txnJobsList = txnAndJobListPair->second;
+          delete txnJobsList;
+          txn2JobsListMap_.erase(txnAndJobListPair->first);
+        }
+        weightedTxnsQueue_.pop(); // remove the empty
+        if (weightedTxnsQueue_.empty())
+        {
+          break;
+        }
+        weightedTxn = weightedTxnsQueue_.top();
+        txnAndJobListPair = txn2JobsListMap_.find(weightedTxn.second);
+      }
+
+      if (weightedTxnsQueue_.empty())
+      {
+        newJob.wait(lk); // might need a lock here
+        continue;
+      }
+
+      // We have non-empty jobsList at this point.
+      // Remove the txn from a queue first to add it later
+      weightedTxnsQueue_.pop(); // remove the empty
```

Comment:
Can you please clarify the comment `remove the empty`, seems like at this point `joblist` for the `top` transaction is not empty.

Reply:
The comment is misguiding so I will remove it.